### PR TITLE
[LD-138] migration: migrate all the dependencies from hilt to koin

### PR DESCRIPTION
**Pull request by trOnk12**
# Pull-Request
## Description
 - Migrate all the dependencies from hilt to koin.
### Why?
- Becase we want to expirment with KMM, we are not able to use android specific libraries.

### What?
- Use koin instead of hilt

## Links to related issues
- https://loudius.atlassian.net/browse/LD-138

## Demo
- Not applicable 

- Run the app should not crash :) 

## Documentation
https://github.com/InsertKoinIO/koin

## Checklist

- [x] Functionality is covered by unit tests
- [x] Functionality is covered by integration tests
- [x] I've updated PR description with relevant information
- [x] I've done self code review
- [x] I've manually tested if the code and app works

### CODE REVIEW

File: app/src/test/java/com/appunite/loudius/di/CheckModulesTest.kt [hunk 1]
```diff
@@ -0,0 +1,45 @@
+package com.appunite.loudius.di
```
Comment (trOnk12) [inline app/src/test/java/com/appunite/loudius/di/CheckModulesTest.kt]: added a CheckModulesTest, which tests the dependencies if all are correctly injected

File: app/build.gradle [hunk 2]
```diff
@@ -118,6 +112,13 @@ dependencies {
     //gson
     implementation 'com.google.code.gson:gson:2.10.1'
 
+    // koin
+    def koinVersion = "3.4.3"
```
Comment (nowakweronika) [inline app/build.gradle]: Suggestion: I would move it into project `build.gradle`. We already have one defined version here.

File: app/src/main/java/com/appunite/loudius/LoudiusApplication.kt [hunk 3]
```diff
@@ -17,7 +17,35 @@
...
+import com.appunite.loudius.di.networkModule
+import com.appunite.loudius.di.repositoryModule
+import com.appunite.loudius.di.serviceModule
+import com.appunite.loudius.di.viewModelModule
 import dagger.hilt.android.HiltAndroidApp
+import org.koin.android.ext.koin.androidContext
+import org.koin.core.context.GlobalContext.startKoin
+import org.koin.dsl.module
 
 @HiltAndroidApp
```
Comment (nowakweronika) [inline app/src/main/java/com/appunite/loudius/LoudiusApplication.kt]: Question: Why do we still have hilt annotation?

File: app/src/main/java/com/appunite/loudius/MainActivity.kt [hunk 4]
```diff
@@ -38,11 +37,12 @@ import com.appunite.loudius.ui.login.LoginScreen
 import com.appunite.loudius.ui.pullrequests.PullRequestsScreen
 import com.appunite.loudius.ui.reviewers.ReviewersScreen
 import dagger.hilt.android.AndroidEntryPoint
+import org.koin.androidx.viewmodel.ext.android.viewModel
 
 @AndroidEntryPoint
```
Comment (nowakweronika) [inline app/src/main/java/com/appunite/loudius/MainActivity.kt]: Question: We do we still have `AndroidEntryPoint` annotation?

File: app/src/main/java/com/appunite/loudius/di/APIQualifiers.kt [hunk 5]
```diff
@@ -21,7 +21,3 @@ import javax.inject.Qualifier
 @Qualifier
 @Retention(AnnotationRetention.RUNTIME)
 annotation class BaseAPI
-
-@Qualifier
-@Retention(AnnotationRetention.RUNTIME)
-annotation class AuthAPI
```
Comment (nowakweronika) [inline app/src/main/java/com/appunite/loudius/di/APIQualifiers.kt:27]: Question: Why do we only removed this annotation class?

File: app/src/main/java/com/appunite/loudius/di/RepositoryModule.kt [hunk 6]
```diff
@@ -28,23 +28,12 @@ import dagger.Module
 import dagger.Provides
 import dagger.hilt.InstallIn
 import dagger.hilt.components.SingletonComponent
```
Comment (nowakweronika) [inline app/src/main/java/com/appunite/loudius/di/RepositoryModule.kt]: Core: I think we should remove dagger imports.

File: app/src/main/java/com/appunite/loudius/domain/repository/AuthRepository.kt [hunk 7]
```diff
@@ -19,8 +19,6 @@ package com.appunite.loudius.domain.repository
 import com.appunite.loudius.domain.store.UserLocalDataSource
 import com.appunite.loudius.network.datasource.AuthDataSource
 import com.appunite.loudius.network.model.AccessToken
-import javax.inject.Inject
-import javax.inject.Singleton
```
Comment (nowakweronika) [inline app/src/main/java/com/appunite/loudius/domain/repository/AuthRepository.kt:23]: Question: Is it really building in the correct way? We have annotations below and here are deleted imports ;)

File: app/src/main/java/com/appunite/loudius/network/datasource/AuthDataSource.kt [hunk 8]
```diff
@@ -21,8 +21,6 @@ import com.appunite.loudius.network.model.AccessToken
 import com.appunite.loudius.network.model.AccessTokenResponse
 import com.appunite.loudius.network.services.AuthService
 import com.appunite.loudius.network.utils.WebException
-import javax.inject.Inject
-import javax.inject.Singleton
```
Comment (nowakweronika) [inline app/src/main/java/com/appunite/loudius/network/datasource/AuthDataSource.kt:25]: Question: Is it really building in the correct way? We have annotations below and here are deleted imports ;)

File: app/src/main/java/com/appunite/loudius/network/datasource/PullRequestsDataSourceImpl.kt [hunk 9]
```diff
@@ -21,8 +21,6 @@ import com.appunite.loudius.network.model.RequestedReviewersResponse
 import com.appunite.loudius.network.model.Review
 import com.appunite.loudius.network.model.request.NotifyRequestBody
 import com.appunite.loudius.network.services.PullRequestsService
-import javax.inject.Inject
-import javax.inject.Singleton
```
Comment (nowakweronika) [inline app/src/main/java/com/appunite/loudius/network/datasource/PullRequestsDataSourceImpl.kt:25]: Question: Is it really building in the correct way? We have annotations below and here are deleted imports ;)

File: app/src/main/java/com/appunite/loudius/network/datasource/UserDataSource.kt [hunk 10]
```diff
@@ -18,8 +18,6 @@ package com.appunite.loudius.network.datasource
 
 import com.appunite.loudius.network.model.User
 import com.appunite.loudius.network.services.UserService
-import javax.inject.Inject
-import javax.inject.Singleton
```
Comment (nowakweronika) [inline app/src/main/java/com/appunite/loudius/network/datasource/UserDataSource.kt:22]: Question: Is it really building in the correct way? We have annotations below and here are deleted imports ;)

File: app/src/main/java/com/appunite/loudius/ui/authenticating/AuthenticatingScreen.kt [hunk 11]
```diff
@@ -25,10 +25,11 @@ import com.appunite.loudius.R
 import com.appunite.loudius.components.components.LoudiusFullScreenError
 import com.appunite.loudius.components.components.LoudiusLoadingIndicator
 import com.appunite.loudius.components.theme.LoudiusTheme
+import org.koin.androidx.compose.koinViewModel
 
 @Composable
 fun AuthenticatingScreen(
-    viewModel: AuthenticatingViewModel = hiltViewModel(),
```
Comment (nowakweronika) [inline app/src/main/java/com/appunite/loudius/ui/authenticating/AuthenticatingScreen.kt:32]: Core: We still have import with `hiltViewModel`.

File: app/src/main/java/com/appunite/loudius/ui/login/LoginScreen.kt [hunk 12]
```diff
@@ -43,11 +43,12 @@ import com.appunite.loudius.components.components.LoudiusOutlinedButtonIcon
 import com.appunite.loudius.components.components.LoudiusOutlinedButtonStyle
 import com.appunite.loudius.components.components.LoudiusText
 import com.appunite.loudius.components.components.LoudiusTextStyle
+import org.koin.androidx.compose.koinViewModel
 import com.appunite.loudius.components.R as componentsR
 
 @Composable
 fun LoginScreen(
-    viewModel: LoginScreenViewModel = hiltViewModel(),
```
Comment (nowakweronika) [inline app/src/main/java/com/appunite/loudius/ui/login/LoginScreen.kt]: Core: We still have import with `hiltViewModel`.

File: app/src/main/java/com/appunite/loudius/ui/reviewers/ReviewersScreen.kt [hunk 13]
```diff
@@ -61,11 +61,12 @@ import com.appunite.loudius.components.components.LoudiusTopAppBar
 import com.appunite.loudius.components.theme.LoudiusTheme
 import com.appunite.loudius.ui.reviewers.ReviewersSnackbarType.FAILURE
 import com.appunite.loudius.ui.reviewers.ReviewersSnackbarType.SUCCESS
+import org.koin.androidx.compose.koinViewModel
 import com.appunite.loudius.components.R as componentsR
 
 @Composable
 fun ReviewersScreen(
-    viewModel: ReviewersViewModel = hiltViewModel(),
```
Comment (nowakweronika) [inline app/src/main/java/com/appunite/loudius/ui/reviewers/ReviewersScreen.kt:69]: Core: We still have import with `hiltViewModel`.

File: app/src/main/java/com/appunite/loudius/ui/reviewers/ReviewersViewModel.kt [hunk 14]
```diff
@@ -61,8 +60,7 @@ enum class ReviewersSnackbarType {
     SUCCESS, FAILURE
 }
 
-@HiltViewModel
```
Comment (nowakweronika) [inline app/src/main/java/com/appunite/loudius/ui/reviewers/ReviewersViewModel.kt:66]: Core: We still have import with `hiltViewModel`.

File: app/src/test/java/com/appunite/loudius/di/CheckModulesTest.kt (hunk 1 above)
Comment (nowakweronika) [reply app/src/test/java/com/appunite/loudius/di/CheckModulesTest.kt]: Praise: More useful tests❤️

File: app/src/main/java/com/appunite/loudius/LoudiusApplication.kt (hunk 3 above)
Comment (trOnk12) [reply app/src/main/java/com/appunite/loudius/LoudiusApplication.kt]: you are correct ;s not needed anymore

File: app/build.gradle
Comment (nowakweronika) [inline app/build.gradle:1]: Question: Why do we have dependencies for `hilt` in project?

File: app/src/main/java/com/appunite/loudius/di/NetworkModule.kt
Comment (nowakweronika) [inline app/src/main/java/com/appunite/loudius/di/NetworkModule.kt:1]: Core: Here are the dependencies for `Hilt`.

**Review by nowakweronika — COMMENTED** [review]
Why is this draft? Do you want to do something more in this branch?

File: app/build.gradle
Comment (trOnk12) [reply app/build.gradle:1]: They were there for insturmented test I will remove it in the next PR

File: app/src/main/java/com/appunite/loudius/di/NetworkModule.kt
Comment (trOnk12) [reply app/src/main/java/com/appunite/loudius/di/NetworkModule.kt:1]: forgot it ;s thanks for the reminder

**Comment by trOnk12** [comment]
> Why is this draft? Do you want to do something more in this branch?

I made it draft because I wanted to make sure we do not merge it to your branch :P I though it will be good idea if you go first with your changes to the develop and I follow you after that

File: app/src/androidTest/java/com/appunite/loudius/TestApplication.kt [hunk 15]
```diff
@@ -0,0 +1,31 @@
+package com.appunite.loudius
```
Comment (jacek-marchwicki) [inline app/src/androidTest/java/com/appunite/loudius/TestApplication.kt]: chore: not license header in the file

File: app/src/androidTest/java/com/appunite/loudius/util/InstrumentationTestRunner.kt [hunk 16]
```diff
@@ -0,0 +1,16 @@
+package com.appunite.loudius.util
```
Comment (jacek-marchwicki) [inline app/src/androidTest/java/com/appunite/loudius/util/InstrumentationTestRunner.kt]: chore: not license header in the file

File: app/src/androidTest/java/com/appunite/loudius/TestApplication.kt [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+import com.appunite.loudius.di.repositoryModule
+import com.appunite.loudius.di.serviceModule
+import com.appunite.loudius.di.viewModelModule
+import org.koin.android.ext.koin.androidContext
+import org.koin.core.context.GlobalContext.startKoin
+
+class TestApplication : Application() {
+    override fun onCreate() {
+        super.onCreate()
+        startKoin {
```
Comment (jacek-marchwicki) [inline app/src/androidTest/java/com/appunite/loudius/TestApplication.kt]: chore: With this `TestApplication` implementation we could have bugs in `LoudiusApplication` that are not covered by tests. I'm strongly recommend re-using `LoudiusApplication` in tests, just replace injects that are necessary in tests.

File: app/src/androidTest/java/com/appunite/loudius/LoginScreenTest.kt [hunk 18]
```diff
@@ -67,12 +56,6 @@ class LoginScreenTest {
     @JvmField
     val screenshotTestRule = ScreenshotTestRule()
 
-    @Before
-    fun setUp() {
-        hiltRule.inject()
-    }
-
-    @BindValue
```
Comment (jacek-marchwicki) [inline app/src/androidTest/java/com/appunite/loudius/LoginScreenTest.kt]: bug: You've removed `BindValue` from tests, this means that `GithubHelperModule` is not replaced with this implementation. This definitely should break our UI tests, but somehow they are not executed. This should be also investigated why UI tests are not executed.

**Review by jacek-marchwicki — CHANGES_REQUESTED** [review]
Sorry... I haven't finished checking the whole PR. This is just a partial review.

File: app/src/androidTest/java/com/appunite/loudius/LoginScreenTest.kt (hunk 18 above)
Comment (trOnk12) [reply app/src/androidTest/java/com/appunite/loudius/LoginScreenTest.kt]: thanks for noticing !

File: app/src/androidTest/java/com/appunite/loudius/TestApplication.kt (hunk 17 above)
Comment (trOnk12) [reply app/src/androidTest/java/com/appunite/loudius/TestApplication.kt]: good idea

File: app/src/androidTest/java/com/appunite/loudius/LoginScreenTest.kt (hunk 18 above)
Comment (trOnk12) [reply app/src/androidTest/java/com/appunite/loudius/LoginScreenTest.kt]: or actually from what I understand this was kind of removing the GitHelperModule which provided GithubHelper so that we could mock it, but now since we are not using Hilt for injection, we do not need to do it

Comment (trOnk12) [reply app/src/androidTest/java/com/appunite/loudius/LoginScreenTest.kt]: nevermind I see what you mean now, trying to fix it ...

File: app/src/main/java/com/appunite/loudius/network/utils/AuthFailureHandler.kt [hunk 19]
```diff
@@ -21,11 +21,8 @@ import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.flow.MutableSharedFlow
 import kotlinx.coroutines.flow.SharedFlow
 import kotlinx.coroutines.launch
-import javax.inject.Inject
-import javax.inject.Singleton
 
-@Singleton
-class AuthFailureHandler @Inject constructor(
+class AuthFailureHandler(
```
Comment (jacek-marchwicki) [inline app/src/main/java/com/appunite/loudius/network/utils/AuthFailureHandler.kt:25]: **question**: I wonder if this still be singleton, cause is not yet annotated by `@Signelton` and it's not declared in a module as singleton.

File: app/src/test/java/com/appunite/loudius/di/CheckModulesTest.kt [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+import org.koin.dsl.koinApplication
+import org.koin.test.KoinTest
+import org.koin.test.check.checkModules
+import org.koin.test.mock.MockProvider
+
+@ExtendWith(MainDispatcherExtension::class)
+class CheckModulesTest : KoinTest {
+
+    @Test
+    fun verifyKoinApp() {
```
Comment (jacek-marchwicki) [inline app/src/test/java/com/appunite/loudius/di/CheckModulesTest.kt:35]: **suggestion**: When do you think this test might fail? 

AFAIK this test might be much more useful if it would test logic that is used in `LoudiusApplication`

File: app/src/test/java/com/appunite/loudius/di/CheckModulesTest.kt [hunk 21]
```diff
@@ -0,0 +1,42 @@
+package com.appunite.loudius.di
```
Comment (jacek-marchwicki) [inline app/src/test/java/com/appunite/loudius/di/CheckModulesTest.kt:17]: **chore**: There is no license header in this file.

File: app/src/main/java/com/appunite/loudius/di/ViewModelModule.kt [hunk 22]
```diff
@@ -0,0 +1,17 @@
+package com.appunite.loudius.di
```
Comment (jacek-marchwicki) [inline app/src/main/java/com/appunite/loudius/di/ViewModelModule.kt:17]: **chore**: There is no license header in this file

**Review by jacek-marchwicki — CHANGES_REQUESTED** [review]
Some minor changes are required to merge this PR.

File: app/src/test/java/com/appunite/loudius/di/CheckModulesTest.kt (hunk 20 above)
Comment (trOnk12) [reply app/src/test/java/com/appunite/loudius/di/CheckModulesTest.kt:35]: this test failes whenever there are some definitions when that cannot be resolved, kind of like Hilt verifies it during compile time.

Comment (trOnk12) [reply app/src/test/java/com/appunite/loudius/di/CheckModulesTest.kt:35]: + good suggestion 🎖️

File: app/src/androidTest/java/com/appunite/loudius/LoginScreenTest.kt (hunk 18 above)
Comment (trOnk12) [reply app/src/androidTest/java/com/appunite/loudius/LoginScreenTest.kt]: should be good now 💯

File: app/src/main/java/com/appunite/loudius/network/utils/AuthFailureHandler.kt (hunk 19 above)
Comment (trOnk12) [reply app/src/main/java/com/appunite/loudius/network/utils/AuthFailureHandler.kt:25]: it is 🤔  inside NetworkModule 

singleOf(::AuthFailureHandler)

File: app/src/test/java/com/appunite/loudius/di/CheckModulesTest.kt (hunk 20 above)
Comment (trOnk12) [reply app/src/test/java/com/appunite/loudius/di/CheckModulesTest.kt:35]: used the appModules reference to the appmodule excluding viewmodel module as it is causing headache, because of tests complaining about not returning stuff for sharedprefs for some keys that the values are retrieved when the ViewModel is initialized.

 I do not think we are loosing much not checking for ViewModels, as mostly ViewModel is last item in the dependency three, which should have "bigger blocks" being injected, so all the most important validation is happening before viewmodels injection, let me know if that is okay with you and if you agree with it or not...

File: app/src/androidTest/java/com/appunite/loudius/TestApplication.kt (hunk 17 above)
Comment (jacek-marchwicki) [reply app/src/androidTest/java/com/appunite/loudius/TestApplication.kt]: This class still exists.

File: app/src/test/java/com/appunite/loudius/di/CheckModulesTest.kt [hunk 23]
```diff
@@ -0,0 +1,54 @@
...
+class CheckModulesTest : KoinTest {
+
+    @Test
+    fun verifyKoinApp() {
+        val mockContext = mockkClass(Context::class)
+        val mockSharedPref = mockkClass(SharedPreferences::class)
+
+        every { mockContext.getSharedPreferences(any(), any()) } returns mockSharedPref
+
+        MockProvider.register { mockkClass(it) }
```
Comment (jacek-marchwicki) [inline app/src/test/java/com/appunite/loudius/di/CheckModulesTest.kt]: suggestion: you don't need `MockProvider.register { mockkClass(it) }`. It can be removed.

File: app/src/main/java/com/appunite/loudius/di/NetworkModule.kt [hunk 24]
```diff
@@ -17,94 +17,66 @@
...
-            url(baseUrl)
-        }
-        install(ContentNegotiation) {
-            register(ContentType.Application.Json, GsonConverter(gson))
+    single<HttpClient>(named("base")) {
+        HttpClient(OkHttp) {
+            expectSuccess = true
+            engine {
+                addInterceptor(TestInterceptor)
+                addInterceptor(get<HttpLoggingInterceptor>())
```
Comment (nowakweronika) [inline app/src/main/java/com/appunite/loudius/di/NetworkModule.kt:66]: Question: Why did you delete `authInterceptor` and `authFailureHandler` ?

File: app/src/main/java/com/appunite/loudius/network/datasource/UserDataSource.kt (hunk 10 above)
Comment (trOnk12) [reply app/src/main/java/com/appunite/loudius/network/datasource/UserDataSource.kt:22]: should be fixed now :)

File: app/src/main/java/com/appunite/loudius/domain/repository/AuthRepository.kt (hunk 7 above)
Comment (trOnk12) [reply app/src/main/java/com/appunite/loudius/domain/repository/AuthRepository.kt:23]: should be good now :)

File: app/src/main/java/com/appunite/loudius/network/datasource/AuthDataSource.kt (hunk 8 above)
Comment (trOnk12) [reply app/src/main/java/com/appunite/loudius/network/datasource/AuthDataSource.kt:25]: should be good now :)

File: app/src/main/java/com/appunite/loudius/di/NetworkModule.kt (hunk 24 above)
Comment (trOnk12) [reply app/src/main/java/com/appunite/loudius/di/NetworkModule.kt:66]: thanks for the sharp eye 🦅 resvoled

File: app/src/main/java/com/appunite/loudius/LoudiusApplication.kt [hunk 25]
```diff
@@ -17,7 +17,45 @@
...
+        serviceModule,
+        repositoryModule,
+        githubHelperModule,
+        dispatcherModule
+    )
+}
+
+class LoudiusApplication : Application() {
+    override fun onCreate() {
+        println("OnCreate")
```
Comment (jacek-marchwicki) [inline app/src/main/java/com/appunite/loudius/LoudiusApplication.kt]: **suggestion:** remove this println

File: app/src/test/java/com/appunite/loudius/ActivitySetupTest.kt [hunk 26]
```diff
@@ -17,36 +17,28 @@
...
 import org.junit.Before
-import org.junit.Rule
 import org.junit.Test
 import org.junit.jupiter.api.DisplayName
 import org.junit.runner.RunWith
 import org.robolectric.annotation.Config
 
 @RunWith(AndroidJUnit4::class)
-@Config(sdk = [Build.VERSION_CODES.Q], application = HiltTestApplication::class)
+@Config(sdk = [Build.VERSION_CODES.Q], application = LoudiusApplication::class)
```
Comment (jacek-marchwicki) [inline app/src/test/java/com/appunite/loudius/ActivitySetupTest.kt]: **suggestion:** AFAIK you can skip specifing application, because it is read from the AndroidManifest.xml file: `, application = LoudiusApplication::class`

File: gradle.properties [hunk 27]
```diff
@@ -21,3 +21,4 @@ kotlin.code.style=official
 # resources declared in the library itself and none from the library's dependencies,
 # thereby reducing the size of the R class for that library
 android.nonTransitiveRClass=true
+android.defaults.buildfeatures.buildconfig=true
```
Comment (jacek-marchwicki) [inline gradle.properties:24]: **question:** why `biuldFeature {buildConfig : true}` is moved from from gradle files to gradle.properties? Is there any benefit of the new approach?

**Review by jacek-marchwicki — APPROVED** [review]

File: gradle.properties (hunk 27 above)
Comment (trOnk12) [reply gradle.properties:24]: it does not have to be defined per each module I think it is something we want to have anyways ?
